onnxconverter_common/auto_mixed_precision_model_path.py

```diff
@@ -0,0 +1,250 @@
+    final_nodes = try_to_convert_to_valid_fp16_model(**kwargs)
+
+    print(".final convert.")
+    kwargs["node_block_list"] = final_nodes
+    kwargs["return_model"] = True
+    valid, model = run_attempt(**kwargs)
+    if not valid:
+        raise ValueError("validation failed for final fp16 model")
+    print("Final model validated successfully.")
+
+    clean_output_folder(**kwargs)
+
+
+def try_to_convert_to_valid_fp16_model(**kwargs):
+    node_names = kwargs.get('node_block_list')
+
+    segments = SegmentList(node_names)
+    i = 0
+    while segments.get_largest() is not None:
+        seg = segments.get_largest()
+        nodes_to_try = segments.get_nodes(seg)
+        i += 1
+        print("Running attempt %d excluding conversion of %s nodes" % (i, len(nodes_to_try)))
+        kwargs["node_block_list"] = nodes_to_try
+        if run_attempt(**kwargs):
+            seg.good = True
+            print("Attempt succeeded.")
+        else:
+            print("Attempt failed.")
+            if seg.size == 1:
+                seg.bad = True
+            else:
+                seg.split()
+        print("segments=", segments)
+    print("Done! these nodes will keep float32 type:", segments.get_nodes())
+
+    return segments.get_nodes()
+
+
+def generate_temp_filename(target_model_path):
+    target_model_folder = os.path.dirname(target_model_path)
+    if not os.path.exists(target_model_folder):
+        os.mkdir(target_model_folder)
+    tensor_filename = str(uuid.uuid1())
+    onnx_filename = os.path.join(target_model_folder, tensor_filename + ".onnx")
+    return onnx_filename, tensor_filename + ".data"
+
+
+def copy_fp32_model(**kwargs):
+    source_model_path = kwargs.get('source_model_path')
+    input_feed = kwargs.get('input_feed')
+    providers = kwargs.get('providers')
+    tmp_model32_path = kwargs.get("tmp_model32_path")
+    tmp_model32_tensor_name = kwargs.get("tmp_model32_tensor_name")
+
+    model_32 = onnx.load(source_model_path)
+    save_model(model_32, tmp_model32_path, location=tmp_model32_tensor_name)
+
+    print("infer_shape_path for", tmp_model32_path, tmp_model32_tensor_name)
+    shape_inference.infer_shapes_path(tmp_model32_path)
+    model_32 = onnx.load(tmp_model32_path)
+
+    output_32 = inference(tmp_model32_path, input_feed, providers=providers)
+
+    kwargs["res1"] = output_32
+    kwargs["res2"] = output_32
+    if not validate(**kwargs):
+        raise ValueError("validation failed for fp32 model")
+
+    return model_32, output_32
+
+
+def validate(**kwargs):
+    validate_fn = kwargs.get("validate_fn")
+    rtol = kwargs.get("rtol")
+    res1 = kwargs.get("res1")
+    res2 = kwargs.get("res2")
+
+    if validate_fn is not None and not validate_fn(res1, res2):
+        return False
+    if rtol is not None:
+        for r1, r2 in zip(res1, res2):
+            if not np.allclose(r1, r2, rtol):
+                return False
+    return True
+
+
+def run_attempt(**kwargs):
```

Comment:
Let's change this name to a meaningful one.

Reply:
resovled